fix(core): Fix decimal pipe floating point formatting bug

injae-kim opened the pull request:
## PR Type
- [x] Bugfix


## What is the current behavior?
```
Input (string) : "123456789.123456789"
Format : 0.0-9
Output :   123,456,789.12345679 // 7(8)9 -> 8 is lost..
Expected : 123,456,789.123456789
```
https://github.com/angular/angular/blob/d315e2c4fa178dfbd41bc25259605bb999fa302e/packages/common/src/pipes/number_pipe.ts#L104-L105

- Currently, `decimal_pipe` parse string to number first and then apply formating
  - ⚠️ When parse `'123456789.123456789'` string to number, some digit on fraction part is lost due to javascript floating point limitation
- Issue Number: https://github.com/angular/angular/issues/47809
  - Reproduce: https://stackblitz.com/edit/angular-ivy-ycarro?file=src/app/app.component.html


## What is the new behavior?
```js
        const formatStrs = [integerPart, fractionPart]
            .map(part => strToNumber(part))
            .map(num => formatNumber(num, locale, digitsInfo));
```
- To format string to number on javascript well, apply formatting on integer/fraction part separately and then concat, return it.

## Does this PR introduce a breaking change?

- [x] No

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,22 @@ export class DecimalPipe implements PipeTransform {
     locale = locale || this._locale;
 
     try {
+      if (typeof value === 'string' && value.includes('.')) {
```
Nit: We should probably add a comment to indicate what particular case we're handling in the block.

injae-kim replied:
oh~ nice suggestion. I'll address your review and fix lint too. thanks!

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,22 @@ export class DecimalPipe implements PipeTransform {
+        const integerStr = formatNumber(strToNumber(integerPart), locale, digitsInfo).split('.')[0];
+        const fractionStr = formatNumber(strToNumber(fractionPart), locale, digitsInfo).split('.')[1];
+
+        if (typeof fractionStr === 'undefined') {
```
Looks like this can be simplified. 
```suggestion
        if (fractionStr === undefined) {
```

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,22 @@ export class DecimalPipe implements PipeTransform {
+        if (typeof fractionStr === 'undefined') {
+          return integerStr;
+        }
+        return integerStr + '.' + fractionStr;
```
This should be a template string
```suggestion
        return `${integerStr}.${fractionStr}`;
```

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,22 @@ export class DecimalPipe implements PipeTransform {
+      if (typeof value === 'string' && value.includes('.')) {
+        const parts = value.split('.');
+        if (parts.length != 2) {
+          throw new Error(`${value} is not a number`);
```
We should like factorate this error. `strToNumber` also throws the same error. This will allow us to reduce the bundle size impact.

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,22 @@ export class DecimalPipe implements PipeTransform {
+          throw new Error(`${value} is not a number`);
+        }
+        const integerPart = parts[0];
+        const fractionPart = '0.' + parts[1];
```
```suggestion
        const fractionPart = `0.${parts[1]}`;
```

JeanMeche commented on packages/common/src/pipes/number_pipe.ts:
```diff
@@ -101,6 +101,21 @@ export class DecimalPipe implements PipeTransform {
+      // e.g. '123456789.123456789'
+      if (typeof value === 'string' && value.includes('.')) {
+        const parts = value.split('.');
+        if (parts.length != 2) {
```
The condition doesn't cover every non valid string, we should likely rely on `strToNumber` to throw.

injae-kim replied:
```
      if (typeof value === 'string' && value.includes('.') && strToNumber(value)) {
        const parts = value.split('.');
        const integerPart = parts[0];
        const fractionPart = `0.${parts[1]}`;
        ..
```
Hmm you mean calling `strToNumber(value)` inside of `if(..)`?
Or can you give some example code? 🙇

JeanMeche replied:
We can either call `strToNumber()` or probably better extract the portion we'd like to reuse in a function : 

```
function isStringNumber(value: number|string): boolean {
  return typeof value === 'string' && !isNaN(Number(value) - parseFloat(value))
}
```

injae-kim replied:
aha~ I understood. `isStringNumber` looks much better! thanks for nice suggestion 👍

injae-kim committed:
fix(core): Fix decimal pipe floating point formatting bug

fix(core): Fix decimal pipe floating point formatting bug
Address `JeanMeche`'s comments
Extract duplicated codes to isStringNumber()

injae-kim commented:
I need this feature and ready to update PR, so waiting your awesome reviews~ thanks! 😃

injae-kim committed:
Merge branch 'main' into 47809-fix-decimal-pipe-floating-point

injae-kim committed:
Update number_pipe_spec.ts
